### PR TITLE
Implement apply for Phase

### PR DESCRIPTION
**Pull request by msohcw**
## Summary
* Implement Phase.apply, State.winner

## Tests
* Added simple integration tests for MoveCommand
* Added tests for State.winner

### CODE REVIEW

File: santorini-common/src/objects/state.rs [hunk 1]
```diff
@@ -50,10 +51,23 @@ impl State {
...
+    pub fn winner(&self) -> Option<Player> {
+        for space in self.board.iter().flatten() {
+            if let Some(worker) = space.worker() {
+                if space.tower().level() == Level::Three {
+                    return Some(worker.player());
+                }
+            }
+        }
+
+        return None;
```
Comment (mingyli) [inline santorini-common/src/objects/state.rs:69]: lint!

File: santorini-common/src/objects/state.rs [hunk 2]
```diff
@@ -50,10 +51,23 @@ impl State {
...
     /// Mutates the game state. Any errors surfaced here are not irrecoverable.
-    fn apply_command(&mut self, command: &dyn Command) -> Result<(), SantoriniError> {
+    pub fn apply_command(&mut self, command: &dyn Command) -> Result<(), SantoriniError> {
         command.execute(self)?;
         Ok(())
     }
+
+    // Returns a winner, assuming that a legal state has exactly one winner
+    pub fn winner(&self) -> Option<Player> {
+        for space in self.board.iter().flatten() {
```
Comment (mingyli) [inline santorini-common/src/objects/state.rs:61]: nice find!

File: santorini-common/src/objects/state.rs [hunk 3]
```diff
@@ -50,10 +51,23 @@ impl State {
...
+            if let Some(worker) = space.worker() {
+                if space.tower().level() == Level::Three {
```
Comment (mingyli) [inline santorini-common/src/objects/state.rs:63]: dang we can't use && with if let?

File: santorini-common/src/objects/state.rs [hunk 4]
```diff
@@ -50,10 +51,23 @@ impl State {
...
+                    return Some(worker.player());
```
Comment (mingyli) [inline santorini-common/src/objects/state.rs:64]: The way you'd structure this code is by using `Iterator::find`. This returns the first instance of an element that satisfies a given predicate, or otherwise None.

File: santorini-common/src/phase.rs [hunk 5]
```diff
@@ -1,7 +1,23 @@
...
+use crate::{command::Command, error::SantoriniError};
 
 #[derive(Debug)]
 pub enum Phase {
     InProgress(State),
     Win(State, Player),
 }
+
+impl Phase {
+    fn apply(self, command: &dyn Command) -> Result<Phase, SantoriniError> {
```
Comment (mingyli) [inline santorini-common/src/phase.rs:11]: Is the goal to eventually use this function in a fold or a try_fold?

File: santorini-common/src/objects/state.rs (hunk 3 above)
Comment (msohcw) [reply santorini-common/src/objects/state.rs:63]: Doesn't look like it. Also I preferred this expression for clarity.

File: santorini-common/src/phase.rs (hunk 5 above)
Comment (msohcw) [reply santorini-common/src/phase.rs:11]: See integration tests.

File: santorini-common/src/objects/state.rs (hunk 4 above)
Comment (msohcw) [reply santorini-common/src/objects/state.rs:64]: I played around with this. `find` doesn't really help the code here. If I were to e.g. find for a "winning" space, I would then need a second bit to destructure the space again to get the player. I used `find_map` instead to do it directly, but then the expression is pretty similar. 

I also considered `filter` but then you do excess work.

File: .github/workflows/rust.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+name: Rust
```
Comment (mingyli) [inline .github/workflows/rust.yml:1]: Hm was this part of this pr? Want to try rebase/merge again?

File: .gitignore [hunk 7]
```diff
@@ -1 +1,7 @@
 /target
+
+# Cargo
+Cargo.lock
```
Comment (mingyli) [inline .gitignore:4]: Hm I suppose this is fine. https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html 
Can you do this in a separate pr?

File: santorini-common/examples/repl.rs [hunk 8]
```diff
@@ -15,10 +15,7 @@ fn read_position(message: &str) -> Position {
         let mut entry = String::new();
         io::stdin().read_line(&mut entry)?;
         match *entry.chars().collect::<Vec<char>>().as_slice() {
-            [column, row, ..] => Ok(Position {
-                column: Column::try_from(column)?,
-                row: Row::try_from(row)?,
-            }),
+            [column, row, ..] => Ok(Position(Column::try_from(column)?, Row::try_from(row)?)),
```
Comment (mingyli) [inline santorini-common/examples/repl.rs:18]: Yea try rebasing this

File: santorini-common/src/objects/state.rs (hunk 4 above)
Comment (mingyli) [reply santorini-common/src/objects/state.rs:73]: `find_map` sgtm. You can modify the inner part to use the combinator `space.worker().and_then(Worker::player)` . Monads in action wow!

Comment (mingyli) [reply santorini-common/src/objects/state.rs:73]: A side note is that we try to avoid using `return` as much as possible

**Review by mingyli — COMMENTED** [review]
good stuff. I think this requires a remerge/rebase with master
